### ISSUE TITLE
Workers not being deployed for Nuxt 3 projects

### ISSUE THREAD

**Issue by jadechip** [cloudflare/workers-sdk#3401]
I would really like some clarity regarding the following:

>               (Edit: Forgot to add that I intended to ask @luc122c about this - thanks for any input!)
> 
> I am a bit confused by this issue: Is the intended behaviour that _both_ the native `/functions` folder from Cloudflare Pages _and_ the `/server/api/` folder from Nuxt 3 are supposed to work in parallel. Or will the `/server` folder from Nuxt be "transformed" into the Cloudflare Pages `/functions` folder?
> 
> By setting the build step to `nuxt build && ls functions` and setting the Nitro preset to "cloudflare_pages" I have success with putting my API functions in `/server/api`, and I can see that the functions folder is then created and populated with two files during deployment on Cloudflare Pages:
> 
> ```
> 03:19:44.892 | functions:
> 03:19:44.892 | total 702
> 03:19:44.892 | -rw-r--r-- 1 buildbot nogroup 305165 Feb 26 02:19 [[path]].js
> 03:19:44.893 | -rw-r--r-- 1 buildbot nogroup 412466 Feb 26 02:19 [[path]].js.map
> ```
> 
> However, I cannot get it to work by simply creating the `/functions` folder in my Nuxt project and adding a HelloWorld function as described on https://developers.cloudflare.com/pages/platform/functions/get-started/ - when I deploy, I get this:
> 
> ```
> 03:03:00.842 | Found Functions directory at /functions. Uploading.
> 03:03:01.419 |  
> 03:03:01.523 | ✘ [ERROR] No routes found when building Functions directory: /opt/buildhome/repo/functions
> 03:03:01.523 |  
> 03:03:01.523 |  
> 03:03:01.524 | If you think this is a bug then please create an issue at https://github.com/cloudflare/workers-sdk/issues/new/choose
> 03:03:01.535 | Warning: Wrangler did not find routes when building functions. Skipping.
> ```
> 
> I am mostly curious if I am misunderstanding how this is supposed to work. Thanks in advance!

_Originally posted by @cjsinnbeck in https://github.com/cloudflare/workers-sdk/issues/1859#issuecomment-1445253000_

**Comment by jadechip** [comment]
For more context, I have a Nuxt 3 project deployed to Cloudflare Pages. 
I also have a `/functions` directory with a Cloudflare Worker but it doesn't appear in the build logs and is not getting deployed. I am using the cloudflare_pages preset: 
```
  nitro: {
    preset: "cloudflare_pages",
    output: {
      dir: path.join(__dirname, 'output')
    }
  },
```
Please advise.

**Comment by petebacondarwin** [comment]
Hi @jadechip - please would you be able to provide a reproduction that I could look at. It is hard to tell what the problem is from the given information. The error suggests that there are no valid route files in the `functions` folder.

**Comment by martijnn** [comment]
I'm also facing this issue.

I'm using Nuxt + Nuxt content to deploy my site to pages. I'm not using any functions folder within my project.

The workaround I've been using is making use of Wrangler to push my code, which is mentioned in this comment: https://github.com/cloudflare/workers-sdk/issues/2475#issuecomment-1377355111

**Comment by penalosa** [comment]
@jadechip we haven't heard from you in a while, so I'm going to close this issue for now. If you're still running into issues feel free to comment with more details and we can investigate further!

**Comment by bllchmbrs** [comment]
I think I am seeing this issue as well. Here is my setup.

1. `"nuxt": "^3.6.5"`
2. `"@nuxt/content": "^2.7.2"`
3. a `/functions` folder that is just [the functions tutorial described by cloudflare docs](https://developers.cloudflare.com/pages/platform/functions/get-started/).

I set the 
```
  nitro: {
    preset: 'cloudflare-pages',
  },
 ```
in my nuxt config and have successfully deploy the site.

if I run `npx wrangler pages dev -- npm run dev` I will be able to 'query' my cloudflare pages function in the browser on my local machine.

However, when I perform the deploy (currently via github). The cloudflare pages function doesn't show up. I get a 404 (from nuxt).

I'm super confused about what I should do here and don't know how to proceed.

Here is [a repo](https://github.com/bllchmbrs/cf-worker-issue) that reproduces this issue, vanilla install of everything. Given that code, this route: https://cf-worker-issue.pages.dev/helloworld should work on cloudflare, but it doesn't.

**Comment by dario-piotrowicz** [comment]
Hi @bllchmbrs I actually think that what you're experiencing is the expected behavior (although I can appreciate that it might be confusing)

In your reproduction locally if you run
```
NITRO_PRESET=cloudflare-pages npm run build
```
you can see that nuxt generates an `_worker.js` file:
![Screenshot 2023-08-02 at 16 55 03](https://github.com/cloudflare/workers-sdk/assets/61631103/5f6de01d-38c4-4115-a86a-148b290ddba9)

That means that Nuxt builds the Pages application in [advanced mode](https://developers.cloudflare.com/pages/platform/functions/advanced-mode/)

Advanced mode and `functions` are two different incompatible things and in advanced mode the `functions` directory just gets ignored:
![Screenshot 2023-08-02 at 16 56 28](https://github.com/cloudflare/workers-sdk/assets/61631103/9968c948-260f-4b1a-9866-7c1d6b4262a3)

So I do think that your `functions` directory getting ignored is expected and just how things work on Pages, the correct thing to do when using Nuxt is just not use the `functions` directory and do everything using Nuxt and its [server handlers](https://nuxt.com/docs/guide/directory-structure/server) they provide access to all the Cloudflare bindings, cf object and ctx object so they should have all you need, is there something that you think can't be done there and needs to be done via Pages Functions?

___

<details>

<summary>helloworld</summary>

The `functions` directory is completely ignored so that's why when you go to  https://cf-worker-issue.pages.dev/helloworld you just end up getting served the Nuxt application (if in your application you were to set a `helloworld` route I'd assume you'd get that)

</details>

<details>

<summary>local development</summary>

`npx wrangler pages dev -- <command>` is used to spin up a worker and provide a way for it to get its static assets, it does so by proxy requests to the server you spin up by the command (`npm run dev` in your case), on top of it if handles the `functions` dir (+ other stuff I'd imagine, I am not really 100% familiar with the Pages inner workings)

so it's not surprising that that does work in your case because there you don't specify the `_worker.js` file anywhere, so `functions` work as you'd expect them to and the rest of the requests goes to your (nexi) dev server.


`npx wrangler pages dev -- <command>` isn't really representative of what really happens in production, to get a faithful representation you should instead do: `npx wrangler pages dev dist/`

(
as suggested by the build script:
![Screenshot 2023-08-02 at 17 16 58](https://github.com/cloudflare/workers-sdk/assets/61631103/f55c0697-da6a-4c99-a429-1d3490683e8e)
)

There you do get a faithful representation of what happens in production and there you can again see that the `functions` directory gets ignored:
![Screenshot 2023-08-02 at 17 20 21](https://github.com/cloudflare/workers-sdk/assets/61631103/c1dbed7c-8c98-480f-a487-e4deaf519bdf)

</details>

**Comment by dario-piotrowicz** [comment]
So yeah sorry for the very long message 😅 (I hope that it helped 🙏)

TLDR; @bllchmbrs I think that the behaviour you're seeing is expected, that also applied to the original issue created by @jadechip , the solution is simply not to use `functions` since they are not supported in advanced mode but just fully rely on Nuxt (I think it should handle everything you'd need it to, I'd be very keen to know what can't be handled there that can in Functions)

**Comment by dario-piotrowicz** [comment]
I however have no idea of @martijnn's issue... maybe Nuxt was doing something with functions at some point and was causing the issue... but if I understand correctly that's no longer a problem right?

**Comment by bllchmbrs** [comment]
Great, 🙇‍♂️ !

I will give this a try. I guess I figured that it wasn't _advanced_ so I didn't dive too deeply in there. My own ignorance. My hope is that there are other folks like me that can find this post and succeed! 💪

thanks again and i will let you know if I experience any issues with this.

One thing that is ... bug-ish... is that wrangler shouldn't "work" with the route. The discrepancy between local and cloudflare makes it feel distinct. Even a wrangler warning saying 
> WARNING: Wrangler found a `functions` directory and a `_worker.js` directory. Please see x-documentation or you may see strange results.

That could go a long way, but I could be the only one that doesn't understand this stuff, so I leave it to you for prioritization. :)

**Comment by dario-piotrowicz** [comment]
@bllchmbrs my pleasure 🙂 

yeah it can happen to everyone, I don't even think that Nuxt clearly states/documents that they use advanced mode, so it's easy to miss I'd imagine

I hope that you'll be able to deploy and work on your Nuxt app without issues 😄 (if you've still got some me and the Cloudflare team is always here to help 🙂)

___

Anyways yes I agree that a warning would be helpful, the issue with that is that when you run your `wrangler pages dev -- npm run dev` you could never have even created the `_worker.js`, so there wouldn't be anything to warn you about (also even if there was an `_worker.js` file, wrangler wouldn't know where to look for it, as it is not in the current directory, so it'd have to look for it or something)
(I think the thing is that we should discourage using the `-- command` version of `wrangler pages dev` as it doesn't produce results that properly represents what then happens in production)

The warning could anyways still be useful when you manually deploy your app or in the pages CI (although less noticeable there), I'll bring this to the team 🙂👍

**Comment by dario-piotrowicz** [comment]
@bllchmbrs I've checked and the team seem to agree that a warning could be helpful, I've created a new issue for that #3701 🙂👍 (hopefully someone will look into it at some point, or I might even give it a shot 😄)
